Fix a bunch of compiler warnings in "make check"

jsquyres opened the pull request:
There were many compiler warnings, making it difficult to read if
there were real issues that needed to be addressed.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

Commit:
Fix a bunch of compiler warnings in "make check"

There were many compiler warnings, making it difficult to read if
there were real issues that needed to be addressed.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

jsquyres commented on test/datatype/reduce_local.c:
```diff
@@ -44,19 +44,12 @@ static op_name_t array_of_ops [] = {
-   ({ __typeof__ (a) _a = (a); \
-       __typeof__ (b) _b = (b); \
-     _a < _b ? _a : _b; })
+#define max(a,b) ((a) > (b) ? (a) : (b))
```
Hey @bosilca -- can you comment on this?  I replaced the prior `min` and `max` macros with the canonical, simpler versions, but I don't know if there was something subtle happening in the old macros (e.g., with making new variable instances via `__typeof__`).  The warning that the compiler was emitting was that ISO C forbids making new variable instances inside expressions.

Do we need something better than the canonical `min`/`max` that I introduced here?

jsquyres replied:
@bwbarrett pointed out to me that the `opal/util/minmax.h` macros are probably a good choice here.

bosilca replied:
Definitively better than the originally proposed replacement.

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/26139a0296a14949b84cd45f2f061e3e

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/32bf8dd20b2c6d67d130adae19301854

ibm-ompi commented:
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/c1d2e04ab921478133c065c997868f85

ibm-ompi commented:
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/4cdba890b56d44a2f2e3c614a5008c57

jjhursey commented:
IBM CI had a bad weekend (sorry about that). We should be all good now.
bot:ibm:pgi:retest
bot:ibm:xl:retest

bosilca commented on test/datatype/partial.c:
```diff
@@ -71,7 +71,7 @@ static void show_neighborhood(double* ptr, int how_many, bool show_hex)
 
 int main( int argc, char* argv[] )
 {
-    opal_datatype_t* vector;
```
This should be opal_datatype_t.

jsquyres replied:
There's a few places where this variable is used where an `(ompi_datatype_t*)` is expected:

```
1044 partial.c:89:55: warning: passing argument 3 of ‘ompi_datatype_create_contigu\
     ous’ from incompatible pointer type [-Wincompatible-pointer-types]
1045      ompi_datatype_create_contiguous(CONT_COUNT, base, &vector);
1046                                                        ^~~~~~~

1051 partial.c:93:24: warning: passing argument 1 of ‘ompi_datatype_dump’ from inc\
     ompatible pointer type [-Wincompatible-pointer-types]
1052      ompi_datatype_dump(vector);
1053                         ^~~~~~

1058 partial.c:97:31: warning: passing argument 1 of ‘opal_datatype_type_extent’ f\
     rom incompatible pointer type [-Wincompatible-pointer-types]
1059      opal_datatype_type_extent(base, &base_extent);
1060                                ^~~~
```

It seemed better to make it `ompi_datatype_t` and then use the `->super` to get to the `opal_datatype_t` where relevant.

Is that incorrect?

bosilca replied:
The ompi_datatype_t is holding additional information that is unused here. But as we use ompi_datatype functions to manipulate it your fix is correct.

bosilca commented on test/datatype/reduce_local.c:
```diff
@@ -44,19 +45,10 @@ static op_name_t array_of_ops [] = {
     { NULL, "MPI_OP_NULL", MPI_OP_NULL }
 };
-static int do_ops[12] = { -1, };  /* index of the ops to do. Size +1 larger than the array_of_ops */
+static int global_do_ops[12] = { -1, };  /* index of the ops to do. Size +1 larger than the array_of_ops */
```
I doubt it is necessary, the shadowing is done in purpose.

jsquyres replied:
The shadowing creates a bunch of warnings, such as:

```
1122 reduce_local.c: In function ‘build_do_ops’:
1123 reduce_local.c:84:34: warning: declaration of ‘do_ops’ shadows a global decla\
     ration [-Wshadow]
1124  build_do_ops( char* optarg, int* do_ops)
1125                              ~~~~~^~~~~~
```

bosilca replied:
I understand what the warning is but I said I do this on purpose. By default all functions use the global variable until I need to specialize their behavior, in which case I add a param shadowing the global variable.

bwbarrett replied:
I'm not a huge fan of that paradigm, as I think it's pretty easy for developers to miss and or get wrong.  I certainly wouldn't want any of the developers in my org to write code like that.  That said, if we are going to intentionally shadow variables, can we at least document that we're doing so intentionally?  I still think it is bad form, and that we should not remove the general compiler warning, so we'll have to figure out how to ignore the warning in this file), but we at least need to tell people it was intentional.

bosilca commented on test/datatype/unpack_ooo.c:
```diff
@@ -151,7 +151,8 @@ static int testcase(ompi_datatype_t * newtype, size_t arr[][2]) {
                 }
                 for (i = 0; 0 != arr[i][0]; i++) {
-                    if( (displ >= arr[i][1]) && (displ <= (arr[i][1] + arr[i][0])) ) {
+                    if( (displ >= (ptrdiff_t) arr[i][1]) &&
```
displ is a ptrdiff_t already because of a compiler complaint, (because it is computed as a difference between two pointers, while in fact is the displacement of a field in a struct).

jsquyres replied:
Here's the compiler warnings it generated without updating the RHS to be `(ptrdiff_t)`:

```
1024 unpack_ooo.c: In function ‘testcase’:
1025 unpack_ooo.c:154:32: warning: comparison of integer expressions of different \
     signedness: ‘ptrdiff_t’ {aka ‘long int’} and ‘size_t’ {aka ‘long unsigned int\
     ’} [-Wsign-compare]
1026                      if( (displ >= arr[i][1]) && (displ <= (arr[i][1] + arr[i\
     ][0])) ) {
1027                                 ^~
1028 unpack_ooo.c:154:56: warning: comparison of integer expressions of different \
     signedness: ‘ptrdiff_t’ {aka ‘long int’} and ‘size_t’ {aka ‘long unsigned int\
     ’} [-Wsign-compare]
1029                      if( (displ >= arr[i][1]) && (displ <= (arr[i][1] + arr[i\
     ][0])) ) {
1030                                                         ^~
```

I don't have an opinion on this one; do you prefer a different fix?

bosilca replied:
Based on a machine understanding of a piece of code. I don't really care.

bosilca commented on test/asm/atomic_cmpset.c:
```diff
@@ -104,71 +104,95 @@ int main(int argc, char *argv[])
 
     /* -- cmpset 32-bit tests -- */
 
-    vol32 = 42, old32 = 42, new32 = 50;
```
This was a perfectly legitimate code. What was the complaint here ?

jsquyres replied:
This is the type of warning the compiler is throwing:

```
649 atomic_cmpset.c:107:15: warning: right-hand operand of comma expression has no effect [-Wunused-value]
650      vol32 = 42, old32 = 42, new32 = 50;
651                ^
```

I didn't dig into the POSIX C standard to see what is exactly going on here; perhaps something to do with the compiler being able to ignore the RHS of a comma operator in an assignment (since the assignment is single value)...?  🤷‍♂️

Commit:
test/datatype/partial: make all output on stderr

Trivial updates to make all the output appear on stderr (vs. just some
of it on stderr and other parts of it on stdout).

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>

gpaulsen commented:
@jsquyres Can you please rebase?   This may be semi-important.

gpaulsen commented:
bot:retest
/azp run
bot:ibm:retest

gpaulsen commented:
bot:ibm:retest

jjhursey commented:
Any reason to hold this PR any longer? (other than it need to be rebased)

gpaulsen commented:
I rebased on tip of `main`(no code change).

gpaulsen committed:
fixing 6 atomic compare calls

Fixing 6 calls that I may have incorrectly addressed in the previous
commit when I rebased to latest main.

Signed-off-by: Geoffrey Paulsen <gpaulsen@us.ibm.com>

jsquyres commented:
If that 3rd commit passes CI properly, you'll squash it appropriately, right?  (minor pet peeve: make a commit on a PR and then make a later commit on the same PR to fix the earlier commit)

bwbarrett reviewed (changes requested):
What's the warning?  The non-sized version is supposed to handle multiple sizes without warning, so if we're emitting warnings here, that's an issue we shouldn't gloss over with this particular patch.

gpaulsen commented:
mellanox CI failed that it couldn't find a definition for the non-sized versions of these, and I searched around and couldn't find those definitions either, so I created the 3rd commit.
Why the other CIs didn't fail, I'm not sure.
@bwbarrett thoughts?

gpaulsen commented:
I git bisected `opal/include/opal/sys/atomic_stdc.h` and couldn't find a definition for `opal_atomic_compare_exchange_strong` anywhere in the past.

bwbarrett commented:
bot:aws:retest

hppritcha commented:
anyone still working on this?

hppritcha commented:
closing.  can be reopened if there's interest.
